zebra-network/src/peer_set/initialize.rs

```diff
@@ -717,15 +713,11 @@ where
                 // Try to connect to a new peer.
                 let _ = demand_tx.try_send(MorePeers);
             }
-            HandshakeConnected { peer_set_change } => {
-                if let Change::Insert(ref addr, _) = peer_set_change {
-                    debug!(candidate.addr = ?addr, "successfully dialed new peer");
-                } else {
-                    unreachable!("unexpected handshake result: all changes should be Insert");
-                }
+            HandshakeConnected { address, client } => {
+                debug!(candidate.addr = ?address, "successfully dialed new peer");
                 // successes are handled by an independent task, so they
                 // shouldn't hang
```

Comment:
I think it's true, because this is inside `crawl_and_dial` which is spawned in a separate task:

https://github.com/ZcashFoundation/zebra/blob/ebc40f8af7a6ca0a375d2effe728a9fc79a64b96/zebra-network/src/peer_set/initialize.rs#L218-L227

Could you confirm if my reasoning is correct?

Should we expand this comment a bit to make it clearer?

Reply:
`candidates.update()` uses the `PeerSet`, and it runs in the crawler task, so this comment is not always true.

We want to spawn each update into a separate task (#1976), just like we spawn each handshake into a separate task. But for now, we just add a short timeout inside `candidates.update()` to break the deadlock.

Once the deadlock is broken, `peerset_tx.send(_)` can populate the `PeerSet` from successful handshakes.

I'm not sure what happens if we delete our whole `AddressBook`, but I can work that out in #1873.

Reply:
```suggestion
                // successes are handled by an independent task,
                // except for `candidates.update`, which has a timeout,
                // so this `send` shouldn't hang
```

Reply:
Done.